testing/upgradetest/versions.go

```diff
@@ -36,33 +41,102 @@ var (
 	Version_8_13_0 = version.NewParsedSemVer(8, 13, 0, "", "")
 )
 
-// GetUpgradableVersions returns the version that the upgradeToVersion can upgrade from.
-func GetUpgradableVersions(ctx context.Context, upgradeToVersion string, currentMajorVersions int, previousMajorVersions int) ([]*version.ParsedSemVer, error) {
-	aac := tools.NewArtifactAPIClient()
-	vList, err := aac.GetVersions(ctx)
+// VersionRequirements is to set requirements for upgradable versions while fetching them.
+//
+// Keep in mind that requirements can overlap. For example 2 previous minors might already
+// cover 2 current majors, so the results would be combined.
+type VersionRequirements struct {
+	UpgradeToVersion string
+	CurrentMajors    int
+	PreviousMajors   int
+	PreviousMinors   int
+	RecentSnapshots  int
+}
+
+const AgentVersionsFilename = ".agent-versions.json"
+
+type AgentVersions struct {
+	// TestVersions contains semver-compliant versions of the agent to run integration tests against.
+	TestVersions []string `json:"testVersions"`
+}
+
+var (
+	agentVersions       *AgentVersions
+	ensureAgentVersions = sync.OnceFunc(setAgentVersions)
+)
+
+func setAgentVersions() {
+	v, err := getAgentVersions()
 	if err != nil {
-		return nil, fmt.Errorf("error retrieving versions from Artifact API: %w", err)
+		panic(err)
 	}
-	if len(vList.Versions) == 0 {
-		return nil, errors.New("retrieved versions list from Artifact API is empty")
+	agentVersions = v
+}
+
+func getAgentVersions() (*AgentVersions, error) {
+	var filePath string
+	dir, err := os.Getwd()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the current directory: %w", err)
+	}
+	for {
+		dir = filepath.Dir(dir)
```

Comment:
How is it better? We're already using the same approach in the integration tests when finding the project root by looking for `go.mod`.

Using `go list` is more complex, since it's an external command. I'd like to keep it as it is.

Reply:
The difference is in code: we don't need to go up one level at the time till the root of the filesystem (potentially in case the file is not the in current directory tree), we could just compose the expected path starting from the repository root. Not blocking anyways